### DISCUSSION TITLE
AsyncSession: Many2Many Relationship Querying

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#5913] [Usage Questions] [answered]
Hello, I was reading about how to query m2m relationships directly from the parent class here: https://docs.sqlalchemy.org/en/13/orm/tutorial.html#building-a-many-to-many-relationship.

More specifically this example:
```
>>> wendy.posts.\
...         filter(BlogPost.keywords.any(keyword='firstpost')).\
...         all()
[BlogPost("Wendy's Blog Post", 'This is a test', <User(name='wendy', fullname='Wendy Williams', nickname='windy')>)]
```

I am unable to reproduce this: `await PARENT_INSTANCE.M2M_COLUMN_REL.filter(...).all()` I am getting the following:

```
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: http://sqlalche.me/e/14/xd2s)
```

Any suggestions on querying child from the parent m2m?

Thanks!

**Comment by zzzeek** [comment, marked as answer]
> Hello, I was reading about how to query m2m relationships directly from the parent class here: https://docs.sqlalchemy.org/en/13/orm/tutorial.html#building-a-many-to-many-relationship.
> 
> ```

First things first.  You're using 1.4 now with 2.0 fully turned on.   Don't use the docs in 1.3:

https://docs.sqlalchemy.org/en/13/ - obsolete

Use the 1.4 docs:

https://docs.sqlalchemy.org/en/14/ - current

additionally, the tutorial you cite is also completely obsolete for 2.0-mode style of working, which includes all of asyncio:

https://docs.sqlalchemy.org/en/14/orm/tutorial.html - obsolete

https://docs.sqlalchemy.org/en/14/core/tutorial.html - obsolete

the tutorial you want to use is:

https://docs.sqlalchemy.org/en/14/tutorial/index.html

Many to many is not specifically in the 1.4 tutorial as of yet, as from an ORM perspective, querying for an m2m relationship is not any different from querying an o2m relationship.  We should probably note this.  but anyway, the official "m2m" document at the moment is:

https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many

which doesn't get much into querying either.   

> More specifically this example:

> >>> wendy.posts.\
...         filter(BlogPost.keywords.any(keyword='firstpost')).\
...         all()
[BlogPost("Wendy's Blog Post", 'This is a test', <User(name='wendy', fullname='Wendy Williams', nickname='windy')>)]

What this API is using is a thing called the dynamic loader strategy, which is documented at https://docs.sqlalchemy.org/en/14/orm/collections.html#dynamic-relationship-loaders .  This section seems to be a bit in the wrong place as it should be under the "loading relationships" top level section, but anyway this is where it is at the moment.

Basically the dynamic loader strategy is not compatible with the asyncio interface unless you use it in a run_sync() block, and it's in fact not really compatible with the overall SQLAlchemy 2.0 style way of working that wants you to have an explicit session, though we have not deprecated it for 2.0 as it is still useful for many people.    the thing that "dynamic" does can be achieved more explicitly using a helper [with_parent](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=with_parent#sqlalchemy.orm.with_parent) which I demonstrate below.

Also, the .all() method that the dynamic loader refers towards emits a query.   If you're working in asyncio, you can't call any method in SQLAlchemy that would emit a query that isn't an "async" method (unless you use the run_sync() block), because blocking IO is not allowed.    That is, you can only use the methods that are in the [asyncio modules themselves](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html) and you cannot use other kinds of SQL execution systems that are documented elsewhere in SQLAlchemy.   As SQLAlchemy is now publishing two APIs that per the design of Python asyncio are fundamentally incompatible, you need to stay within one API or the other.


> Any suggestions on querying child from the parent m2m?

Now that we're through all of that, you are looking for a helper to emit a certain kind of SELECT statement, one where you have an ORM object and you want to SELECT for objects which it relates towards using a relationship.  This helper is available as the [with_parent](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=with_parent#sqlalchemy.orm.with_parent) helper.

I did a litle practice writing this as asyncio (which is new to me too) so here's a demo:
```python

import asyncio

from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import with_parent

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)
    bs = relationship("B", secondary="atob")


atob = Table(
    "atob",
    Base.metadata,
    Column("aid", ForeignKey("a.id"), primary_key=True),
    Column("bid", ForeignKey("b.id"), primary_key=True),
)


class B(Base):
    __tablename__ = "b"
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)


e = create_async_engine(
    "postgresql+asyncpg://scott:tiger@localhost/test", echo=True
)


async def main():
    async with e.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(e) as s:

        s.add_all(
            [
                A(data="a1", bs=[B(data="a1b1"), B(data="a1b2")]),
                A(data="a2", bs=[B(data="a2b1"), B(data="a2b2")]),
            ]
        )

        await s.commit()

        a1 = (
            (await s.execute(select(A).where(A.data == "a2")))
            .scalars()
            .first()
        )

        bs_with_a = select(B).where(with_parent(a1, A.bs))

        print((await s.execute(bs_with_a)).scalars().all())


asyncio.run(main())
```

the query at the end that does the with_parent looks like:

```sql
SELECT b.id, b.a_id, b.data 
FROM b, atob AS atob_1 
WHERE ? = atob_1.aid AND b.id = atob_1.bid

```

**Reply by zikphil** [reply]
Following-up on that, how would I go about adding an element to a many2many relationship field?

`await orm_object.m2m_relationship.append(reminder)` 

also throws the `greenlet_spawn has not been called; can't call await_() here.` error because it is trying to lazy-load the relationship.

Do I need to manually build the relationship using the Table object?

**Reply by zzzeek** [reply]
that would indicate that the relationship is not loaded.  you need to use eager loading so that it's loaded and can then be appended towards.

**Reply by zzzeek** [reply]
@CaselIT didn't we have some notion of a nicer error message when lazy loading fails under asyncio?  or did I reject that idea

**Reply by CaselIT** [reply]
we do have a nicer message now, should be in b2 already: https://github.com/sqlalchemy/sqlalchemy/commit/bb846d1650423ce8013feb93b3c8837b4c5498a0

it also links to this https://docs.sqlalchemy.org/en/14/errors.html?highlight=errors#missinggreenlet

**Reply by zikphil** [reply]
This is running on b3. Also this is after eagery loading the relationship.

**Reply by CaselIT** [reply]
can you please post the full stracktrace? do you have a reproducing case?

**Reply by zikphil** [reply]
Ok I think my issue was with something else.

Using the same session I was loading User ID 1 without any `selectinload` then later I was re-loading the same User ID 1 with the the relationship pre-loaded using `selectinload` but since it was already loaded in the session it returned the original object without the relationship loaded.

```
session = new_session()
user = await session.get(User, 1)
new_user = await session.get(User, 1, options=[selectinload(User.m2m_relationship)])
print(new_user.m2m_relationship) --> Raised sqlalchemy.exc.MissingGreenlet
```

**Reply by CaselIT** [reply]
I think you can use [populate_existing](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session#sqlalchemy.orm.Session.get.params.populate_existing) to force reloading the object. I guess it would use the loader options then

**Reply by connebs** [reply]
Related to this, is there any sort of convenience method in asyncio-land for loading in a relationship without reloading the object? Or for an object that has just been created, not fetched from the DB (and therefore where solutions like `selectinload()` are not applicable)?

For example with `User` and `Post` models:

```py
# create new post
post = Post(user_id=34, content="hello world")

# get associated user from DB, and stick onto `post` object
await session.load_relationship(post.user)

# use relation in subsequent code without triggering IO issues
if post.user.logged_in:
    ...
```

Thoughts? Does it exist in SQLA? If not, can it? Seems pretty useful to me, and honestly I like this approach even better than the automatic lazy-loading from before asyncio support, after all: explicit is better than implicit!

**Reply by zzzeek** [reply]
there's some methods that allow you to load relationships from a transient object but we consider this to be an antipattern, however it's there if you look at https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.enable_relationship_loading .    -

Another way would be to merge your Post into the session, if you can provide its primary key value.  you can create `Post(id=<primary key>)`, then convert it to "detached" using [make_transient_to_detached](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.make_transient_to_detached) then merge() it into the session without emitting any SQL by passing load=False: https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.merge .    then you can load from it.

to accomplish the lazy loading under asyncio you would use the "run_sync()" method of AsyncSession.

now after all that, the way that I think is **best practice** is to not even think in terms of the Post object, if you want the User for a certain Post that has a cerain id in it you can use with_parent:

   user = session.execute(select(User).where(with_parent(Post(user_id=34), Post.user)).scalar_one()
